### ISSUE TITLE
'event_time' in PAX has a time offset compared to the rundb start time of 7 hours

### ISSUE THREAD

**Issue by sanderbreur** [XENON1T/pax#400]
There is a -7 hour offset on the event time in PAX compared to the event time at LNGS for data processed at Midway in Chicago. (The time difference between LNGS and chicago is 7 hours ). Asking the UTC time of an event reduces this to -2 hours offset. 

When loading a dataset and doing:

Here example = 1640

Get the time from the run doc:

```
run_docs[1640]['start'].timestamp:
datetime.datetime(2016, 7, 24, 11, 54, 10)
```

Get the time from the first event:

```
datetime.datetime.fromtimestamp(data['event_time'].min() /units.s)
datetime.datetime(2016, 7, 24, 4, 54, 10, 21547)
```

Get it from UTC:

```
datetime.datetime.utcfromtimestamp(data['event_time'].min() /units.s)
datetime.datetime(2016, 7, 24, 9, 54, 10, 21547)
```

When using paxer to plot:

`paxer --config XENON1T --plot_to_dir test_plot --input /project/lgrandi/xenon1t/raw/160724_1154`
It gives: 04:54:10 UTC

**Comment by tunnell** [comment]
This is part of issue:

https://docs.python.org/3.4/library/datetime.html#datetime.datetime.timestamp

**Comment by coderdj** [comment]
If it helps I'm writing a script using pax data right now and it looks like datetimes are always UTC minus two hours. Is it possible the timezone correction from LNGS to UTC is being erroneously applied when it's not needed? Or applied twice?

**Comment by tunnell** [comment]
There's a two hour shift in event times due to what I link above.  Should be fixed in event builder.

**Comment by tunnell** [comment]
f862b3210d0a6b3d78c834354f673b07b02b9050

**Comment by JelleAalbers** [comment]
Did f862b3210d0a6b3d78c834354f673b07b02b9050 fix it, or is this still present? From the issue it looks like there are multiple offsets, has anyone managed to disentangle them?

**Comment by adambrown1** [comment]
The timestamp in the processed data is correct but is a UTC timestamp and the python default is to use local time. For example:
```
import pytz
tz = pytz.timezone('UTC')
data1 = hax.minitrees.load('160906_0523')
print("Default: ", datetime.datetime.fromtimestamp(data1.iloc[0]['event_time'] / 10**9))
print("UTC: ", datetime.datetime.fromtimestamp(data1.iloc[0]['event_time'] / 10**9, tz=tz))

Default:  2016-09-06 00:23:41.397240
UTC:  2016-09-06 05:23:41.397240+00:00
```

Pull request #466 makes this explicitly done in Plotting.py for the event waveforms generated by paxer

**Comment by JelleAalbers** [comment]
Thanks @adambrown1! If the timestamp is correct all the way up to the processed data, and now it is fixed in the display, that should take care of our timezone problems. Let's reopen this in case a problem reemerges.

**Comment by molinario** [comment]
We found a problem in MV processed data. The time of the first event, variable event.start_time, appears to be 1 hour earlier than the startrun (once I have both of them converted in UTC). 
To be more clear. The start run of the MV is UTC time, and it is reported in the run's name and it matches with the start time of the run I read in the database.
While if I take event.start_time, which is unix time and convert in UTC, it is 1 hour earlier.

So I found that this problem was NOT present until MV run 161031_0446
It shows up starting from MV run 161103_1524

**Comment by JelleAalbers** [comment]
@molinario how exactly did you determine that there was an offset? The timestamps of events are stored as ns since the UTC epoch, without timezone information. Python's routines to 'convert' times to UTC assume you're feeding in naive datetimes generated in the machine's local time, so if you feed them UTC timestamps you get wrong results.

Do you still get an offset if you feed the unconverted timestamp (event.start_time) to e.g. www.epochconverter.com ?

**Comment by molinario** [comment]
@JelleAalbers The way I found this offset is exactly the one you write in the last part of your comment. I took the raw integer timestamp with event.start_time, I divide by 1e9, and I feed to a unix time to UTC converter, namely www.unixtimestamp.com. As an example, MV run 161124_0030 started on 24/11/2016 at 12:30 am (UTC). The raw integer timestamp you get for the first event is the following: 1479943827999992820, which in seconds is 1479943828. If I feed to the unix time converter I find 23/11/2016 at 23:30:28 pm (UTC).

**Comment by feigaodm** [comment]
Is this fixed @sanderbreur ?

**Comment by sanderbreur** [comment]
I do not know, but I will check and dive back into the timing hell.

Op ma 17 jul. 2017 om 20:42 schreef Fei Gao <notifications@github.com>:

**Comment by sanderbreur** [comment]
Quickly double checked for processed TPC data again:

Event from:

'170202_1747'

Shows :

datetime.datetime(2017, 2, 2, 17, 47, 55, 855698)

So that looks good. (using
datetime.datetime.utcfromtimestamp(data.event_time.values[0]/1e9))


But @Fei this looks like a MV problem. @Andrea is this still a problem?




Op ma 17 jul. 2017 om 21:05 schreef Sander Breur <sanderbreur@gmail.com>:

**Comment by molinario** [comment]
@sanderbreur It looks like the problem is still there. I checked for example MV run 170714_0226 processed with pax_v6.6.6 and I considered the variable event.start_time. For the first event, it results to be 1499995595000002760. I divide it by 1e9, and so I get 1499995595. This unix time corresponds to 2017-07-14T01:26:35, exactly 1 hour before the true value. 

Of course I am managing this when producing MV minitrees and checking for coincidences with TPC, but still there is something wrong.

**Comment by sanderbreur** [comment]
Ok, @molinario thanks for checking. So where do you think the problem lies? 

@tunnell or @JelleAalbers any idea from the TPC DAQ point of view where this problem could be in the MV DAQ?

**Comment by JelleAalbers** [comment]
Do you still have this in the muon veto data? If so, could it be you have an older version of the event builder to create the events, or an older version of pax for processing? I haven't heard of any timing problems on the TPC data stream, and we've even had a summer/winter time transition :-)

**Comment by molinario** [comment]
Yes, the problem is still there. In this moment (winter time in Europe) there is a one hour difference among the actual time at which the MV run started and the time recorded in event.start_time. 

The processing of MV data is currently done (as far as  I understand) with pax v6.8.0. I am trying to figure out if the problem comes from the event-builder. If you have any suggestion, I can follow that too.

**Comment by JelleAalbers** [comment]
Thanks for checking. Could it be that the eventbuilder machine's system clock is not UTC? We switched this for the TPC some time ago.  Of course your machine's timezone shouldn't matter, but if this solves the problem we might know better where to look.

**Comment by molinario** [comment]
With my great surprise, this problem has disappeared by itself after the blackout on February 25th. While in the last runs before power shutdown we can still see the 1 hour time difference, in the runs taken after we restarted the machines this is not happening anymore. 

I also tried to change the timezone on reader5 as a check, from CET to UTC, but this didn't have any effect. 

I really don't know what could have changed due to shutdown.
